tests/functional/olp-cpp-sdk-dataservice-read/VersionedLayerClientPrefetchTest.cpp

```diff
@@ -0,0 +1,162 @@
+#include "ApiDefaultResponses.h"
+#include "MockServerHelper.h"
+#include "ReadDefaultResponses.h"
+#include "Utils.h"
+
+namespace {
+
+const auto kMockServerHost = "localhost";
+const auto kMockServerPort = 1080;
+const auto kTestHrn = "hrn:here:data::olp-here-test:hereos-internal-test";
+
+class VersionedLayerClientPrefetchTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    auto network = olp::client::OlpClientSettingsFactory::
+        CreateDefaultNetworkRequestHandler();
+
+    settings_ = std::make_shared<olp::client::OlpClientSettings>();
+    settings_->network_request_handler = network;
+    // setup proxy
+    settings_->proxy_settings =
+        olp::http::NetworkProxySettings()
+            .WithHostname(kMockServerHost)
+            .WithPort(kMockServerPort)
+            .WithType(olp::http::NetworkProxySettings::Type::HTTP);
+    SetUpMockServer(network);
+  }
+
+  void TearDown() override {
+    auto network = std::move(settings_->network_request_handler);
+    settings_.reset();
+    mock_server_client_.reset();
+  }
+
+  void SetUpMockServer(std::shared_ptr<olp::http::Network> network) {
+    // create client to set mock server expectations
+    olp::client::OlpClientSettings olp_client_settings;
+    olp_client_settings.network_request_handler = network;
+    mock_server_client_ = std::make_shared<mockserver::MockServerHelper>(
+        olp_client_settings, kTestHrn);
+  }
+
+  std::shared_ptr<olp::client::OlpClientSettings> settings_;
+  std::shared_ptr<mockserver::MockServerHelper> mock_server_client_;
+};
+
+TEST_F(VersionedLayerClientPrefetchTest, Prefetch) {
+  olp::client::HRN hrn(kTestHrn);
+
+  constexpr auto kTileId = "5901734";
+  constexpr auto kLayer = "testlayer";
+  constexpr auto kQuadTreeDepth = 4;
+  constexpr auto kVersion = 44;
+
+  const auto root_tile = olp::geo::TileKey::FromHereTile(kTileId);
+  auto client = std::make_unique<olp::dataservice::read::VersionedLayerClient>(
+      hrn, kLayer, boost::none, *settings_);
+  std::vector<std::string> tiles_data;
+  tiles_data.reserve(4);
+
+  {
+    SCOPED_TRACE("Prefetch tiles");
+    const auto tile = root_tile.ChangedLevelTo(12);
+    const auto request = olp::dataservice::read::PrefetchTilesRequest()
+                             .WithTileKeys({tile})
+                             .WithMinLevel(12)
+                             .WithMaxLevel(15);
+    {
+      mock_server_client_->MockLookupResourceApiResponse(
+          mockserver::ApiDefaultResponses::GenerateResourceApisResponse(
+              kTestHrn));
+      mock_server_client_->MockGetVersionResponse(
+          mockserver::ReadDefaultResponses::GenerateVersionResponse(kVersion));
+      mock_server_client_->MockGetResponse(
+          kLayer, root_tile, kVersion,
+          mockserver::ReadDefaultResponses::GenerateQuadTreeResponse(
+              root_tile, kQuadTreeDepth, {12}));
+      const olp::geo::TileKey firstChild = root_tile.ChangedLevelBy(1);
```

Comment:
Test seems wrong for me. `tile` level is 12, generated quad tree has only tiles of level 12, so result of prefetch should be `tile` only. Why we get 4 tiles of level 13 instead?

Reply:
It is valid that we receiving from server not all tiles, but only available. You can see it on old prefetch tests.